functions.md

```diff
@@ -94,80 +385,115 @@ To simplify a bit, you can think of each symbol in CL having (at least) two "cel
-8
-CL-USER> (*my-fun* 5)
-Warning: This function is undefined:
-  *MY-FUN*
-~~~
-
-Note that in the last example the function object returned by `(adder 3)` is stored in the _value cell_ of `*my-fun*` - thus the error message. If we want to be able to use the symbol `*my-fun*` in the car of a compound form, we have to explicitely store something in its _function cell_ (which is normally done for us by the macro [`defun`](http://www.lispworks.com/documentation/HyperSpec/Body/m_defun.htm)):
+If we want to be able to use the symbol `*my-fun*` in the car of a compound form, we have to explicitely store something in its _function cell_ (which is normally done for us by the macro [`defun`](http://www.lispworks.com/documentation/HyperSpec/Body/m_defun.htm)):
 
 ~~~lisp
 ;;; continued from above
 CL-USER> (fboundp '*my-fun*)
 NIL
 CL-USER> (setf (symbol-function '*my-fun*) (adder 3))
-#<Interpreted Function "LAMBDA (N)" {4869FA19}>
+#<CLOSURE (LAMBDA (X) :IN ADDER) {10099A5EFB}>
 CL-USER> (fboundp '*my-fun*)
 T
 CL-USER> (*my-fun* 5)
 8
 ~~~
 
-Now we are ready do define `doubler` as well:
+Read the CLHS section about [form evaluation](http://www.lispworks.com/documentation/HyperSpec/Body/03_aba.htm) for more.
+
+## Closures
+
+Closures allow to capture lexical bindings:
+
+~~~lisp
+(let ((limit 3)
+      (counter -1))
+    (defun my-counter ()
+      (if (< counter limit)
+          (incf counter)
+          (setf counter 0))))
+
+(my-counter)
+0
+(my-counter)
+1
+(my-counter)
+2
+(my-counter)
+3
+(my-counter)
+0
+~~~
+
+Or similarly:
 
 ~~~lisp
-CL-USER> (defun doubler (f)
-    (lambda (x) (funcall f x x)))
-DOUBLER
-CL-USER> (doubler #'+)
-#<Interpreted Function "LAMBDA (F)" {48675791}>
-CL-USER> (doubler '+)
-#<Interpreted Function "LAMBDA (F)" {486761B1}>
-CL-USER> (funcall (doubler #'+) 4)
-8
-CL-USER> (funcall (doubler '+) 4)
-8
-CL-USER> (defparameter *my-plus* '+)
-*MY-PLUS*
-CL-USER> (funcall (doubler *my-plus*) 4)
-8
-CL-USER> (defparameter *my-fun* (doubler '+))
-*MY-FUN*
-CL-USER> (funcall *my-fun* 4)
-8
+(defun repeater (n)
+  (let ((counter -1))
+     (lambda ()
+       (if (< counter n)
+         (incf counter)
+         (setf counter 0)))))
+
+(defparameter *my-repeater* (repeater 3))
+;; *MY-REPEATER*
+(funcall *my-repeater*)
+0
+(funcall *my-repeater*)
+1
+(funcall *my-repeater*)
+2
+(funcall *my-repeater*)
+3
+(funcall *my-repeater*)
+0
 ~~~
 
-Note that the argument to `funcall` (and `apply`) can either be the function itself, i.e. `#'+`, or a symbol which has the function in its function cell (is fbound to the function), i.e. `'+`.
 
+See more on [Practical Common Lisp](http://www.gigamonkeys.com/book/variables.html).
+
+## `setf` functions
 
-All of the above is _extremely simplified_ - we haven't even mentioned macros, special forms, symbol macros, self-evaluating objects, and lexical environments. Read the CLHS section about [form evaluation](http://www.lispworks.com/documentation/HyperSpec/Body/03_aba.htm) for the real deal.
```

Comment:
I would move this into a Further Reading section, removing the "Extremely simplified" and doing some other wordsmithing.